API/Auth/DlcsBasicAuthenticationHandler.cs

```diff
@@ -0,0 +1,147 @@
+    ///  - Only an admin key or a key belonging to that customer can call these paths.
+    ///  - A customer path is never anonymous.
+    /// 
+    /// Care should be taken not to confuse the customer whose resource is being called with the
+    /// customer making the call.
+    /// When the caller is admin, these can be different.
+    /// The resource might not be associated with any customer.
+    /// </summary>
+    public class DlcsBasicAuthenticationHandler : AuthenticationHandler<BasicAuthenticationOptions>
+    {
+        private readonly ICustomerRepository customerRepository;
+        private readonly DeliveratorApiAuth deliveratorApiAuth; // TODO - will become apiAuth
+
+        public DlcsBasicAuthenticationHandler(
+            IOptionsMonitor<BasicAuthenticationOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            ISystemClock clock,
+            ICustomerRepository customerRepository,
+            DeliveratorApiAuth deliveratorApiAuth)
+            : base(options, logger, encoder, clock)
+        {
+            this.customerRepository = customerRepository;
+            this.deliveratorApiAuth = deliveratorApiAuth;
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            // skip authentication if endpoint has [AllowAnonymous] attribute
+            // Not all API paths require auth...
+            var endpoint = Context.GetEndpoint();
+            if (endpoint?.Metadata?.GetMetadata<IAllowAnonymous>() != null)
+                return AuthenticateResult.NoResult();
+            
+            // ...but any not marked must have the auth header
+            if (!Request.Headers.ContainsKey("Authorization"))
+                return AuthenticateResult.Fail("Missing Authorization Header in request");
+
+            var headerValue = Request.GetAuthHeaderValue(AuthenticationHeaderUtils.BasicScheme);
+            if (headerValue == null)
+            {
+                return AuthenticateResult.Fail("Missing Authorization Header in request");
+            }
+
+            // for a path like /customers/23/queue, the resourceCustomerId is 23.
+            // This isn't necessarily the customer that owns the api key being used on the call!
+            int? resourceCustomerId = null;
+            if (Request.RouteValues.TryGetValue("customerId", out var customerIdRouteVal))
+            {
+                if (int.TryParse(customerIdRouteVal.ToString(), out int result))
+                {
+                    resourceCustomerId = result;
+                }
+            }
+            var authHeader = Encoding.UTF8.GetString(Convert.FromBase64String(headerValue.Parameter));
+            string[] keyAndSecret = authHeader.Split(':');
+            var apiCaller = new {Key = keyAndSecret[0], Secret = keyAndSecret[1]};
+            var customerForKey = await customerRepository.GetCustomerForKey(apiCaller.Key, resourceCustomerId);

```

Comment:
Replace lines 79-81 with a single call? `var apiCaller = await GetApiCaller(headerValue.Parameter)`?

_Really_ slim chance of it happening but this method could have error handling to deal with a invalid auth header (e.g. something without ":" as that would throw an exception as-is).